lib/services/auth/auth_service.dart

```diff
@@ -0,0 +1,129 @@
+import 'dart:convert';
+
+import 'package:shared_preferences/shared_preferences.dart';
+import 'package:survey/gen/configs.gen.dart';
+import 'package:survey/services/locator/locator_service.dart';
+import 'package:survey/models/auth_token_info.dart';
+import 'package:object_mapper/object_mapper.dart';
+import 'package:survey/models/user_info.dart';
+import 'package:survey/services/api/api_service.dart';
+import 'package:survey/services/http/http_service.dart';
+import 'package:survey/services/user/user_service.dart';
+
+part 'params/auth_login_params.dart';
+
+abstract class AuthService {
+  static const loginEndPoint = "/oauth/token";
+  static const logoutEndPoint = "/oauth/revoke";
+  static const preferenceTokenKey = "auth_service_preference_token";
+
+  bool get isAuthenticated;
+
+  String? get token;
+
+  UserInfo? get user;
+
+  Future<void> login({
+    required AuthLoginParams params,
+  });
+
+  Future<void> logout();
+
+  Future<void> attempt();
+
+  Future<void> fetchUser();
+
+  Future<void> attemptAndFetchUser();
+}
+
+class AuthServiceImpl implements AuthService {
+  final ApiService _apiService = locator.get();
+  final UserService _userService = locator.get();
+  String? _token;
+  UserInfo? _user;
+
+  @override
+  bool get isAuthenticated => _token != null;
+
+  @override
+  String? get token => _token;
+
+  @override
+  UserInfo? get user => _user;
+
+  @override
+  Future<void> login({
+    required AuthLoginParams params,
+  }) async {
+    final AuthTokenInfo token = await _apiService.call(
+      method: HttpMethod.post,
+      endPoint: AuthService.loginEndPoint,
+      params: params,
+    );
+    await _storeToken(token);
+    await attemptAndFetchUser();
+  }
+
+  @override
+  Future<void> logout() async {
+    try {
+      await _apiService.call(
+        method: HttpMethod.post,
+        endPoint: AuthService.logoutEndPoint,
+      );
+    } on Exception catch (_) {}
+
+    await _clearToken();
+    _token = null;
+    _user = null;
+  }
+
+  @override
+  Future<void> attempt() async {
+    final token = await _getToken();
+    if (token == null) {
+      _token = null;
+      _user = null;
+      return;
+    }
+
+    _token = token.accessToken;
+    _apiService.configureGlobalToken(_token);
```

Comment:
I'm a bit confusing here, `token`, `_token`, `_getToken` 😳  😓 
For renaming suggestion below, I think it's easier to read

Reply:
@manh-t 
> I'm a bit confusing here, token, _token, _getToken

What I try to describe here that `token` is like a `token` object, not a string and this object contains `accessToken` and `refreshToken`. 
Maybe `authToken` is better? What do you think?


> For renaming suggestion below, I think it's easier to read

I don't prefer to a suffix `Info` into a variable name, because our common way to define model name is: `User`, `AuthToken`... then their names are just `user` or `authToken`

*Noted: I try to add `Info` into my model's class name to prevent conflict with name of the others*